openfisca_core/entities.py

```diff
@@ -30,92 +37,42 @@ def __repr__(self):
         return "Role({})".format(self.key)
 
 
-class Entity(object):
-    """
-        Represents an entity (e.g. a person, a household, etc.) on which calculations can be run.
-    """
-    key = None
-    plural = None
-    label = None
-    doc = ""
-    is_person = False
-
-    def __init__(self, simulation):
-        self.simulation = simulation
+class Population(object):
+    def __init__(self, entity):
+        self.simulation = None
+        self.entity = entity
         self._holders = {}
         self.count = 0
         self.ids = []
-        self.step_size = 0
 
-    def clone(self, new_simulation):
-        """
-            Returns an entity instance with the same structure, but no variable value set.
-        """
-        new = self.__class__(new_simulation)
-        new_dict = new.__dict__
+    def clone(self, simulation):
+        result = Population(self.entity)
+        result.simulation = simulation
+        result._holders = {variable: holder.clone(result) for (variable, holder) in self._holders.items()}
+        result.count = self.count
+        result.ids = self.ids
+        return result
 
-        for key, value in self.__dict__.items():
-            if key == '_holders':
-                new_dict[key] = {
-                    name: holder.clone(new)
-                    for name, holder in self._holders.items()
-                    }
-            elif key != 'simulation':
-                new_dict[key] = value
+    def empty_array(self):
+        return np.zeros(self.count)
 
-        return new
+    def filled_array(self, value, dtype = None):
+        return np.full(self.count, value, dtype)
 
     def __getattr__(self, attribute):
         projector = get_projector_from_shortcut(self, attribute)
         if not projector:
-            raise AttributeError("Entity {} has no attribute {}".format(self.key, attribute))
+            raise AttributeError("Entity {} has no attribute {}".format(self.entity.key, attribute))
```

Comment:
Why not renaming `Entity` to` Population` here too?

Reply:
Agreed, and for now "Population" is (soon) in the implementor's vocabulary but not (yet) in the user's. I've used a more generic formulation that's also (I hope) slightly more helpful.